### ISSUE TITLE
NetstringReceiver.sendString accepts lists and handles them incorrectly

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3299]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)| @Peaker reported|
|-|-|
|Trac ID|trac#3299|
|Type|defect|
|Created|2008-06-18 18:33:01Z|

It uses str() on the given object (via %s) and sends that, but specifies the length of the non-stringified object.

Either stringify first, and use len() on the string, or disallow non-strings.

Attachments:

* [twisted.protocols.basic.sendString.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/12c/12cfe6ee51fcd5b92f1d7c6a5525bbcb4d120d74/0eca0cbd823f4e748ac84337ba5e9a23c7fb8906.patch) (554 bytes) - added by Imp3 on 2009-08-12 19:38:34Z - 
* [twisted.protocols.basic.sendString_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/12c/12cfe6ee51fcd5b92f1d7c6a5525bbcb4d120d74/cdf19cdd583eec9bd489d93f9b56b2059999215d.patch) (1888 bytes) - added by Imp3 on 2009-08-12 20:39:06Z - Added Testcase
* [sendString3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/12c/12cfe6ee51fcd5b92f1d7c6a5525bbcb4d120d74/3425119f35a9aad9e05dcd373e06773c15525ac6.patch) (2800 bytes) - added by jesstess on 2009-12-16 18:45:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3299 3299
type__defect defect
reporter__Peaker Peaker
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1213813981000000 1213813981000000
changetime__1260993746000000 1260993746000000
version__None None
owner__ 
cc__jesstess cc__Imp3
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Imp3's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Imp3 removed owner|
|-|-|

The patch takes care that the sended data is consistent (according to DJBs Netstring Protocol) by stringifing every object before sending (and calculating the size). I think this is wanted behavior, the method is called sendString so we stringify every non-string object to a string. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to Imp3|
|-|-|

Thanks for working on this.

The addition of a docstring for that method is long overdue. :)  This change needs unit tests to go with it.  There should be tests that verify that the right thing is done when a non-string is passed to `sendString`.

I think it would probably be best if `sendString` didn't accept non-strings, but that's what it has always done in the past (by its use of %s to handle the input), so it should probably keep supporting that case.

**Comment by twisted-trac** [comment]
|<img alt="Imp3's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Imp3 removed owner|
|-|-|

Added a test method for sendString which passes in my working copy and fails on 8.2.0

Yes, an exception would be nicer if non string types get fed to this method, as of right now I realize that noone would have thrown something noneString into this method as it would mess up protocol consistency. NetstringReceiver would disconnect if something like that happens.

So maybe the exception + raise_test would be the better approach?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to Imp3|
|-|-|

Looking better.  Thanks again for working on this.

I think the best way to deal with this would be to have a `DeprecationWarning` for non-`str` types, eventually to be changed to an exception.  It would be great for you to get some practice with Twisted's deprecation tools, since it's a skillset that remains far too obscure.  Feel free to shout for help if you need it.  (If you would rather deal with that separately, feel free to file a separate ticket, as well.  I don't think it's a requirement to get this one merged.)

Here are a few review points:

  1. The test docstring should be phrased a bit more positively.  Tell me what "correct" means in this context.  For example: "`L{basic.NetstringReceiver.sendString} will send objects that are not strings, by sending their string representation according to `str()`."  You don't need to say "test that" or "verify that"; it's obvious that's what you're doing from the fact that it's a test :).  Just say what you're asserting.
  1. I also think the test should have a different name.  The name "`test_sendString`" doesn't say anything about non-string types (in fact it strongly implies the test will be testing sending strings!).  Maybe `test_sendNonString`.
  1. `nonStrings` is longer than 80 characters; the line should be wrapped.
  1. Since `nonStrings` is only used in this one test, I think it should be in the body of the test itself, rather than as a class variable.  The variables you are copying, "`strings`" and "`illegalStrings`", are class variables because they interact with the superclass, `LPTestCaseMixin`.  (If you want to refactor this to deal with all the other string sending/receiving classes, that would be great for another ticket, but first things first.)
  1. `sendString`'s docstring should start with a newline, like this:
```
#!py
def sendString(self, data): 
    """
    A method for sending a Netstring.
    ...
    """
```

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_6"></a>|@jesstess removed owner|
|-|-|

sendString3.patch addresses glyph's points and adds a DeprecationWarning.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jesstess|
|-|-|

Looks good, except there should also be a test for the warning itself, using `TestCase.flushWarnings` or one of the other warning-related test helpers.  The warning should also include the version of Twisted in which the deprecation was/will be introduced.  You can use 10.0 as a placeholder.  If that turns out to be wrong, the release manager will fix it.

Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_8"></a>|@jesstess removed owner|
|-|-|

Patch updated to include version number and test for the warning.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @jesstess|
|-|-|

Looks great, please add a news file and apply! :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

Woops, one other thing: the closing triple quote on the sendString docstring is misindented.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_11"></a>|@jesstess set status to `closed`|
|-|-|

(In [27789]) Apply sendString3.patch

Authors: Imp3, jesstess
Reviewers: exarkun, glyph
Fixes: [#3299](https://github.com/twisted/twisted/issues/3299)

Stringify non-string data to sendString before calculating the length
so that the calculated length is equal to the actual length of the
transported data. Also add a DeprecationWarning for this non-string
support.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
